Change Group block default variation in inserter.

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->


## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Fixes #43760.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Removes the effect updating layout in the block edit, adds a constrained layout variation as default and removes flow layout variation from inserter. 

Fixtures are updated because removing the effect caused a re-positioning of block attribute output.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Add a Group block in a post;
2. Check that it has constrained layout enabled;
3. Check that legacy Group blocks with different layout types are not affected.
4. In the site editor, add a template part to a template;
5. Add a Group block inside the template part;
6. Duplicate the template part and check site editor isn't frozen,

## Screenshots or screencast <!-- if applicable -->

tellthemachines committed:
Change Group block default variation in inserter.

talldan commented on packages/block-library/src/group/variations.js:
```diff
@@ -10,12 +10,22 @@ const variations = [
+	},
+	{
+		name: 'group-flow',
+		title: __( 'Group' ),
+		description: __( 'Gather blocks in a container.' ),
+		attributes: { layout: { type: 'default' } },
+		scope: [ 'block' ],
 		isActive: ( blockAttributes ) =>
 			! blockAttributes.layout ||
 			! blockAttributes.layout?.type ||
-			blockAttributes.layout?.type === 'default' ||
-			blockAttributes.layout?.type === 'constrained',
+			blockAttributes.layout?.type === 'default',
 		icon: group,
```
Interesting. I didn't realise this approach had been taken before and there was already a group variation.

I wonder if this old variation can be removed? I'm not sure what a `scope` of `block` does, but any matching blocks would (or should) fallback to the normal block type, which would be fine.

tellthemachines replied:
Good question, it seems to work well enough if I remove it. Let's try that.

tellthemachines committed:
Remove unnecessary variation

andrewserong reviewed (commented):
Thanks for pursuing this alternative! It seems to be working pretty well so far, which is promising.

I noticed that the update to the variations causes the selector at the top right of the inspector controls to toggle off Group when switching to the flow / default layout type:

![2022-09-15 13 41 35](https://user-images.githubusercontent.com/14988353/190309736-51e2a527-9492-469f-b08d-e0e3e40e0d57.gif)

I believe this is where the active block variation check happens in the `BlockVariationTransforms` component, in case it helps: https://github.com/WordPress/gutenberg/blob/33d84b036592a5bf31af05b7710f3b2b14163dc4/packages/block-editor/src/components/block-variation-transforms/index.js#L118

Are there any other places where the active variation check is being used like that? From a quick search for usage of `getActiveBlockVariation`, that appears to be the main one that would affect the Group block as far as I can tell 🤔

talldan commented:
@tellthemachines I think there are going to be a lot of snapshots to update. You may be able to cherry-pick 946e555e221737abd922146004dc42c1f96fe7fe from my branch.

talldan commented:
> Are there any other places where the active variation check is being used like that? From a quick search for usage of getActiveBlockVariation, that appears to be the main one that would affect the Group block as far as I can tell 🤔

Possibly the `isActive` for the new variation should be `true` whenever the `isActive` for row and stack are `false`. What do you think?

Might be able to refactor it to something like: `isActive: ( ...args ) => ! isRowActive( ...args ) && ! isStackActive( ...args )`.

andrewserong commented:
> What do you think?

Oh, I like that. Yes, sounds like a good way to ensure that it's a catch-all for everything that isn't Row or Stack 👍

tellthemachines commented:
> I noticed that the update to the variations causes the selector at the top right of the inspector controls to toggle off Group when switching to the flow / default layout type:

My fault, I removed [these lines](https://github.com/WordPress/gutenberg/pull/44176/files#diff-1076d9cb26ab701d1cbb5d6a12a12eae41f28269a67b200890703bd48ef7df2aL17) while experimenting and forgot to re-add them 🤦

tellthemachines committed:
Add back check for default layout Groups

tellthemachines commented:
I'm not sure what should be expected when transforming a block or a multi-selection into a Group. It would be more consistent with the new default behaviour if those Groups also had constrained layout 🤔 (looking at the failed snapshots and wondering if we should consider some of them legitimate failures 😅 )

tellthemachines committed:
Multi-block transforms should also have constrained layout.

tellthemachines committed:
Update failing snapshots.

tellthemachines commented:
Ok I updated the multi-select transform so wrapping other blocks in a Group will create a constrained layout block. That fixed about half the failing snapshots 😅 and I've just updated the other half.

glendaviesnz commented:
This is testing well for me, can add all the group variations and all see to behave as expected, and  site editor no longer freezes when adding template parts with groups.

I also manually ran the group deprecations in the editor and all updated without any errors. Just one question I had was is `"inherit":true` still needed as an attribute when the block has been updated and `"type":"constrained"` added, ie. is `"layout":{"inherit":true,"type":"constrained"}` a valid combination of layout attributes as it doesn't seem like any new blocks end up with that combination, but migrated ones do?

andrewserong reviewed (commented):
This is testing really well for me, too — I didn't run into any issues with any of the legacy or current markup that I tried it with 👍

I spotted one other place where we can Group blocks together, when multiple blocks are selected — should this one also get the constrained layout for consistency with the contextual menu Group item? https://github.com/WordPress/gutenberg/blob/33d84b036592a5bf31af05b7710f3b2b14163dc4/packages/block-editor/src/components/convert-to-group-buttons/toolbar.js#L17

<img width="352" alt="image" src="https://user-images.githubusercontent.com/14988353/190561880-7fe8b0cd-07ba-45fa-916c-51e3f57f0a2a.png">

I personally don't mind it with the default layout type, so don't feel strongly about which type is used when grouping things either way.

tellthemachines commented:
> is "layout":{"inherit":true,"type":"constrained"} a valid combination of layout attributes as it doesn't seem like any new blocks end up with that combination, but migrated ones do?

That's a good question! It comes from #42763 where we add the new layout type to any legacy blocks with either `inherit` or `contentSize`.  I guess the `inherit` attribute could be removed, but it's not breaking anything either so I opted to forgo the extra logic 🤷 

We could always change that though!

glendaviesnz commented:
> We could always change that though!

I don't have a strong opinion on it, but it may be worth tidying up to avoid confusing future us maybe?

tellthemachines committed:
Grouping paragraphs should add constrained layout.

tellthemachines commented:
> I spotted one other place where we can Group blocks together, when multiple blocks are selected — should this one also get the constrained layout for consistency with the contextual menu Group item?

Good catch! I think it makes sense to do so, and have updated the toolbar accordingly.

andrewserong reviewed (approved):
Great work @tellthemachines, this is testing well for me, and I haven't been able to find any issues with it. Legacy block markup appears to be rendered / updated correctly, and I haven't run into any errors.

I thought for a moment that there was an issue with justification alignments, but it turns out it was a typo that's been resolved in `trunk` in https://github.com/WordPress/gutenberg/pull/44179.

This seems like a hopefully fairly safe change given that the fixtures only needed to be updated to shift the position of the block attributes.

LGTM! 🎉

tellthemachines commented:
Thanks for the reviews folks!

ockham commented:
I just cherry-picked this PR to the wp/6.1 branch to get it included in the next release: 83937b3318
